CMSgov/easi-app
# Easi 3054/it gov task states

# [EASI-3054](https://jiraent.cms.gov/browse/EASI-3054)

## Changes and Description

- Introduce a state for each step that needs to be persisted in SQL and GQL.
    - requestFormState
    - draftBusinessCaseState
    - finalBusinessCaseState
    - decisionState
- Introduce a computed property for columns that already determine state and don't need another column persisted
    - grtMeetingState
     - grbMeetingState
     
   I didn't add another column for the meeting states because it effectively denormalizes the data. We can say if a meeting was scheduled or not by the presence of a date. There was mention of what should happen if a meeting gets rescheduled, but this should fit in with the current paradigm. If additional state does need to be saved for the future, we can implement this in the ticket to implement the trb meeting state.


I also chose to store one shared type for the form state. Each form had the exact same state, and it is unlikely to change in the future. However, if it does change, we can implement a unique state at that time. ( I did have a previous version in the commit history that used unique enums)


## How to test this change
1. Create a system intake
```gql
mutation createSystemIntake{
  
  createSystemIntake(input:{
      requestType :MAJOR_CHANGES
      requester: {
          name: "Mr. Fun Person"
      }
    } )
  
  {
    id   
  }
  
  
}
```
2. Get the intake and verify the states are as expected

```gql
query systemIntake{
  
  systemIntake(id:"{{systemIntakeID}}")
  {
    requestFormState
    draftBusinessCaseState
    grtMeetingState
    finalBusinessCaseState
    grbMeetingState
    decisionState


      itGovTaskStatuses {      
        intakeFormStatus
        feedbackFromInitialReviewStatus
        decisionAndNextStepsStatus
        bizCaseDraftStatus
        grtMeetingStatus
        bizCaseFinalStatus
        grbMeetingStatus

      }    
    
  }
  
  
}
```
3. The response should look like this.
```JSON
{
    "data": {
        "systemIntake": {
            "requestFormState": "NOT_STARTED",
            "draftBusinessCaseState": "NOT_STARTED",
            "grtMeetingState": "NOT_SCHEDULED",
            "finalBusinessCaseState": "NOT_STARTED",
            "grbMeetingState": "NOT_SCHEDULED",
            "decisionState": "NO_DECISION",
            "itGovTaskStatuses": {
                "intakeFormStatus": "READY",
                "feedbackFromInitialReviewStatus": "CANT_START",
                "decisionAndNextStepsStatus": "CANT_START",
                "bizCaseDraftStatus": "CANT_START",
                "grtMeetingStatus": "CANT_START",
                "bizCaseFinalStatus": "CANT_START",
                "grbMeetingStatus": "CANT_START"
            }
        }
    }
}
```


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
